### DISCUSSION TITLE
Is tracing intended to be used as I am using it? How can I trace errors chain?

### DISCUSSION THREAD

**Discussion by frederikhors** [tokio-rs/tracing#2329] [Q&A] [answered]
I'm using [async-graphql](https://github.com/async-graphql/async-graphql) and [axum](https://github.com/tokio-rs/axum).

This is a reproduction of the issue: https://github.com/frederikhors/iss-async-graphql-error-handling.

To start:

- `cargo run`

If you open the GraphiQL client at http://localhost:8000, you can use the below query to simulate what I'm trying to understand:

```graphql
mutation {
  mutateWithError
}
```

The backend response is:

```json
{
  "data": null,
  "errors": [
    {
      "message": "I cannot mutate now, sorry!",
      "locations": [
        /*...*/
      ],
      "path": ["mutateWithError"]
    }
  ]
}
```

I like this, but what I don't understand is the tracing part:

```log
2022-09-29T17:01:14.249236Z  INFO async_graphql::graphql:84: close, time.busy: 626µs, time.idle: 14.3µs
  in async_graphql::graphql::parse
  in async_graphql::graphql::request

2022-09-29T17:01:14.252493Z  INFO async_graphql::graphql:108: close, time.busy: 374µs, time.idle: 8.60µs
  in async_graphql::graphql::validation
  in async_graphql::graphql::request

2022-09-29T17:01:14.254592Z  INFO async_graphql::graphql:146: error, error: I cannot mutate now, sorry!
  in async_graphql::graphql::field with path: mutateWithError, parent_type: Mutation, return_type: String!
  in async_graphql::graphql::execute
  in async_graphql::graphql::request

2022-09-29T17:01:14.257389Z  INFO async_graphql::graphql:136: close, time.busy: 2.85ms, time.idle: 30.8µs
  in async_graphql::graphql::field with path: mutateWithError, parent_type: Mutation, return_type: String!
  in async_graphql::graphql::execute
  in async_graphql::graphql::request

2022-09-29T17:01:14.260729Z  INFO async_graphql::graphql:122: close, time.busy: 6.31ms, time.idle: 7.80µs
  in async_graphql::graphql::execute
  in async_graphql::graphql::request

2022-09-29T17:01:14.264606Z  INFO async_graphql::graphql:56: close, time.busy: 16.1ms, time.idle: 22.6µs
  in async_graphql::graphql::request
```

Do you see the `INFO async_graphql::graphql:146: error, error: I cannot mutate now, sorry!`?

1. Why is it an `INFO` event? I would expect it to be an `ERROR` event.

1. And where is the innter error `this is a DB error`?

### Code

The code is very simple:

```rust
pub struct Mutation;

#[Object]
impl Mutation {
    async fn mutate_with_error(&self) -> async_graphql::Result<String> {
        let new_string = mutate_with_error().await?;

        Ok(new_string)
    }
}

async fn mutate_with_error() -> anyhow::Result<String> {
    match can_i_mutate_on_db().await {
        Ok(s) => Ok(s),
        Err(err) => Err(err.context("I cannot mutate now, sorry!")),
    }
}

async fn can_i_mutate_on_db() -> anyhow::Result<String> {
    bail!("this is a DB error!")
}

async fn graphql_handler(
    schema: Extension<Schema<Query, Mutation, EmptySubscription>>,
    req: GraphQLRequest,
) -> GraphQLResponse {
    schema.execute(req.into_inner()).await.into()
}
```

**Comment by hawkw** [comment, marked as answer]
I'm not familiar with the `async-graphql` library, but it looks like the tracing event is being emitted by that library in this case. The event is at the `INFO` level rather than the `ERROR` level because that library's authors chose to record the event at `INFO`: https://github.com/async-graphql/async-graphql/blob/ad05baf79aea55a0827291abf55af5ce7cc8d386/src/extensions/tracing.rs#L146-L148

Similarly, the reason the `tracing` event includes only the top-level error message is because the `async-graphql` authors chose to write 
```rust
error = %err.message(),
```
when emitting the `tracing` event I linked above. That code is saying "record a field named `error` whose value is `err.message()`, formatted using `fmt::Display`. 

If the error type in this code implements `std::error::Error`, it should instead be possible to use the [`tracing::Value` implementation for `dyn Error` trait objects](https://docs.rs/tracing/latest/tracing/trait.Value.html#impl-Value-for-dyn%20Error%20+%20%27static), in which case, the `tracing` subscriber would receive the value as a `&dyn Error + 'static`. This would allow it to record the error chain as well as the top-level error's message, which many `tracing::Subscriber` implementations (such as the `fmt` subscriber you're using) will do.

To summarize, it's possible to record errors the way you want using `tracing`, but the code in `async-graphql` isn't recording the error the way you want to. It's possible that crate's maintainers would accept a pull request and/or issue to change how the error event is recorded. I think all that's necessary should be to change the code I linked above to something like this:
```rust
                tracinglib::error!(target: "async_graphql::graphql",
                                   error = &err
                                   "error");
```
or, if the compiler doesn't like that, you may need to explicitly cast to `dyn Error`:
```rust
                tracinglib::error!(target: "async_graphql::graphql",
                                   error = &err as &(dyn std::error::Error + 'static)
                                   "error");
```

If the maintainers aren't interested in accepting a PR to change how the event is recorded, you could probably use a modified version of their `TracingExtension` that changes how errors are recorded in your own code...but, again, I'm not personally familiar with that crate, so I'm not totally sure if it's possible to implement extensions in user code.

**Reply by frederikhors** [reply]
Eliza,

_what a wonderful answer!
It's one of those answers that a neophyte dreams of having (it's like taking a child by the hand in the dark). THANK YOU!_

I already understood the problem of `info!` instead of `error!` because someone replied to me in the Rust forum and I PR https://github.com/async-graphql/async-graphql/pull/1091.

For the rest I'm opening an issue to discuss it with the async-graphql team.

Thanks again, you were amazing!

**Reply by hawkw** [reply]
Glad I could help! :)

**Reply by frederikhors** [reply]
I tried the code you suggested:

```rust
async fn resolve(
    //...
    next.run(ctx, info)
        .map_err(|err| {
            tracinglib::error!(target: "async_graphql::graphql",
                              error = %err,
                              "error");
            err
        })
        .instrument(span)
        .await
}
```

the event now is an `ERROR` one **but I still can't get my errors chain**:

```log
2022-09-29T19:58:51.309593Z  INFO iss_async_graphql_error_handling:92: Open graphiql at http://localhost:8000/graphiql

2022-09-29T19:59:20.349539Z  INFO async_graphql::graphql:84: close, time.busy: 620µs, time.idle: 15.1µs
  in async_graphql::graphql::parse
  in async_graphql::graphql::request

2022-09-29T19:59:20.352913Z  INFO async_graphql::graphql:108: close, time.busy: 367µs, time.idle: 8.10µs
  in async_graphql::graphql::validation
  in async_graphql::graphql::request

2022-09-29T19:59:20.354956Z ERROR async_graphql::graphql:146: error, error: I cannot mutate now, sorry!
  in async_graphql::graphql::field with path: mutateWithError, parent_type: Mutation, return_type: String!
  in async_graphql::graphql::execute
  in async_graphql::graphql::request

2022-09-29T19:59:20.357894Z  INFO async_graphql::graphql:136: close, time.busy: 3.00ms, time.idle: 28.8µs
  in async_graphql::graphql::field with path: mutateWithError, parent_type: Mutation, return_type: String!
  in async_graphql::graphql::execute
  in async_graphql::graphql::request

2022-09-29T19:59:20.360733Z  INFO async_graphql::graphql:122: close, time.busy: 5.96ms, time.idle: 7.50µs
  in async_graphql::graphql::execute
  in async_graphql::graphql::request

2022-09-29T19:59:20.362585Z  INFO async_graphql::graphql:56: close, time.busy: 13.8ms, time.idle: 21.5µs
  in async_graphql::graphql::request
```

I see nowhere `this is a DB error!`.

**Why?**

**Reply by frederikhors** [reply]
I think this doesn't print because of these lines: https://github.com/async-graphql/async-graphql/blob/master/src/error.rs#L59-L68.

**Reply by hawkw** [reply]
> I tried the code you suggested:
> 
> ```rust
> async fn resolve(
>     //...
>     next.run(ctx, info)
>         .map_err(|err| {
>             tracinglib::error!(target: "async_graphql::graphql",
>                               error = %err,
>                               "error");
>             err
>         })
>         .instrument(span)
>         .await
> }
> ```
> 
> the event now is an `ERROR` one **but I still can't get my errors chain**:
> 
> ```
> ... snip ...
>
> 2022-09-29T19:59:20.354956Z ERROR async_graphql::graphql:146: error, error: I cannot mutate now, sorry!
>   in async_graphql::graphql::field with path: mutateWithError, parent_type: Mutation, return_type: String!
>   in async_graphql::graphql::execute
>   in async_graphql::graphql::request
> 
> ... snip ...
> ```
> 
> I see nowhere `this is a DB error!`.
> 
> **Why?**

You are still recording the error using its `std::fmt::Display` implementation, rather than its `std::error::Error` implementation. The `%` sigil [tells `tracing` to record the error as a `&dyn fmt::Display` trait object](https://docs.rs/tracing/latest/tracing/#recording-fields). If you look at the code I suggested, you'll see that I didn't use the `%` sigil in either suggestion.

When the error is recorded using the `Error` trait, the `tracing` subscriber will format the rest of the error chain.

**Reply by frederikhors** [reply]
Oh!

Now the error is:

```
147 | ...                   error = &err as &(dyn std::error::Error + 'static),
    |                               ^^^^ the trait `StdError` is not implemented for `ServerError`
    |
    = note: required for the cast from `ServerError` to the object type `dyn StdError`
```

Using:

```rust
                tracinglib::error!(target: "async_graphql::graphql",
                                error = &err as &(dyn std::error::Error + 'static),
                                "error");
```

**Reply by hawkw** [reply]
Looking at the documentation for `async-graphql`, it looks like the `ServerError` type [does not implement the `std::error::Error` trait](https://docs.rs/async-graphql/latest/async_graphql/struct.ServerError.html#trait-implementations). So, you cannot record it as an `Error` trait object. I'm actually quite surprised that it doesn't implement `std::error::Error`, but again, I've never actually used that crate before. Perhaps the contributors to the crate have more insight.

I do notice that it has [a number of fields that are publicly exposed, including a source field](https://docs.rs/async-graphql/latest/async_graphql/struct.ServerError.html#fields). So, perhaps you could modify the `tracing` event to manually include those fields.

**Reply by hawkw** [reply]
Sorry I gave you a totally wrong suggestion that you could record this error using `std::error::Error`, by the way --- I assumed that the trait would be implemented, as implementing it for library error types like this is a pretty common practice in the ecosystem, and it's generally the case that public error types will have an `Error` implementation. But again, I've never actually used the `async-graphql` crate. Sorry about that!

**Reply by radix** [reply]
@hawkw Is  `error = &err as &(dyn std::error::Error + 'static)` still the only way to get errors logged with source chains? This is really cumbersome syntax and I wonder if the tracing ecosystem has improved in the past three years to make this a little easier. (unfortunately, I'm still getting a compile error when I pass `error = err` or `error = &err` directly, saying that my error doesn't implement tracing::Value).
